pkg/integration/transaction_test.go

```diff
@@ -105,29 +106,56 @@ func TestTransaction_Rollback(t *testing.T) {
 	require.Equal(t, "defaultdb", res.Rows[0].Values[0].GetS())
 	require.Equal(t, "db1", res.Rows[1].Values[0].GetS())
 
-	tx, err := client.NewTx(context.TODO())
-	require.NoError(t, err)
+	t.Run("double rollback should result in no transaction found error", func(t *testing.T) {
+		tx, err := client.NewTx(context.TODO())
+		require.NoError(t, err)
 
-	err = tx.SQLExec(context.TODO(), `CREATE TABLE table1(
-		id INTEGER,
-		PRIMARY KEY id
+		err = tx.SQLExec(context.TODO(), `CREATE TABLE table1(
+			id INTEGER,
+			PRIMARY KEY id
 		);`, nil)
-	require.NoError(t, err)
+		require.NoError(t, err)
 
-	err = tx.Rollback(context.TODO())
-	require.NoError(t, err)
+		err = tx.Rollback(context.TODO())
+		require.NoError(t, err)
 
-	err = tx.Rollback(context.TODO())
-	require.Error(t, err)
-	require.Equal(t, "no transaction found", err.Error())
+		err = tx.Rollback(context.TODO())
+		require.Error(t, err)
+		require.Equal(t, "no transaction found", err.Error())
+	})
 
-	tx1, err := client.NewTx(context.TODO())
-	require.NoError(t, err)
+	t.Run("rollback should succeed after failed query statement", func(t *testing.T) {
+		tx, err := client.NewTx(context.TODO())
+		require.NoError(t, err)
 
-	res, err = tx1.SQLQuery(context.TODO(), "SELECT * FROM table1", nil)
-	require.Error(t, err)
-	require.Equal(t, "table does not exist (table1)", err.Error())
-	require.Nil(t, res)
+		res, err = tx.SQLQuery(context.TODO(), "SELECT * FROM table1", nil)
+		require.Error(t, err)
+		require.Equal(t, "table does not exist (table1)", err.Error())
+		require.Nil(t, res)
+
+		err = tx.Rollback(context.TODO())
+		require.NoError(t, err)
+	})
+
+	t.Run("rollback should succeed after failed exec statement", func(t *testing.T) {
+		tx, err := client.NewTx(context.TODO())
+		require.NoError(t, err)
+
+		err = tx.SQLExec(context.TODO(), `CREATE TABLE table1(
+			id INTEGER,
+			PRIMARY KEY id
+		);`, nil)
+		require.NoError(t, err)
+
+		err = tx.SQLExec(context.TODO(), `
+			INSERT INTO table1 (id) VALUES (1);
+			INSERT INTO table1 (id) VALUES (1);
+		`, nil)
+		require.EqualError(t, err, "key already exists")
+
+		err = tx.Rollback(context.TODO())
```

Comment:
This fails

Reply:
I think we could even extend this test - after failure in a single statement you should still be able to execute another valid statement and successfully commit the transaction.

Such failed statement in the middle should not invalidate the whole transaction but a single statement.

But then it may be much harder to implement.

What I believe is happening here is that a broken INSERT that violates constraints could make the internal state of the transaction invalid and the simplest way to handle that situation would be to destroy the transaction completely.

A proper fix would be to implement some internal rollback within a single transaction (thus some form of nested transactions).

Reply:
You are correct that a failing `INSERT` results in a closed transaction. This only makes sense when the transaction is created within `execPreparedStmts`. However, when a transaction is provided as a parameter for `execPreparedStmts`, it shouldn't close the transaction.

I am not sure if you want to continue executing statements after the failure of a single statement. In that case you could use the `UPSERT` or `IF NOT EXISTS` statements. What benefit would continuing give?

Reply:
After discussing this with @ldej:

The proposed solution is to add a new flag (e.g. `broken`) into SQLTx class. That flag when set to true would block any operations on Tx data - both reads and writes (`newKeyReader`, `get`, `set`, `existKeyWith`, `commit`) - allowing only the `Cancel` operation. That way we can ensure there are no operations that would rely on the potentially broken state of the transaction.